openssl-dynamic/src/main/c/ocsp.c

```diff
@@ -0,0 +1,221 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "tcn.h"
+#include "ssl_private.h"
+
+// OCSP stapling should be present in OpenSSL as of version 1.0.0
+#ifndef TCN_OCSP_MIN_OPENSSL_VERSION_NUMBER
+#define TCN_OCSP_MIN_OPENSSL_VERSION_NUMBER 0x10002000L
+#endif /* TCN_OCSP_MIN_OPENSSL_VERSION_NUMBER */
+
+#ifndef TCN_NO_OCSP
+#if defined(OPENSSL_NO_OCSP) || (OPENSSL_VERSION_NUMBER < TCN_OCSP_MIN_OPENSSL_VERSION_NUMBER)
+#define TCN_NO_OCSP 1
+#endif
+#endif /* TCN_NO_OCSP */
+
+#if !defined(TCN_NO_OCSP) && !defined(OPENSSL_IS_BORINGSSL)
+
+static const int OCSP_CLIENT_ACK = 1;
+static const int OCSP_SERVER_ACK = SSL_TLSEXT_ERR_OK;
+
+/**
+ * This function is being called from OpenSSL. We do everything in
+ * Java-land and this callback is just a stub that returns the
+ * right values to keep OpenSSL happy.
+ *
+ * The arg that is passed into this function is the pointer for one
+ * of these values:
+ *   OCSP_CLIENT_ACK
+ *   OCSP_SERVER_ACK
+ */
+static int openssl_ocsp_callback(SSL *ssl, void *arg) {
+    return *(const int*)arg;
+}
+
+#endif /* !defined(TCN_NO_OCSP) && !defined(OPENSSL_IS_BORINGSSL) */
+
+/**
+ * Enables OCSP stapling for the given SSLContext
+ */
+TCN_IMPLEMENT_CALL(void, SSLContext, enableOcsp)(TCN_STDARGS, jlong ctx, jboolean client) {
+
+    tcn_ssl_ctxt_t *c = J2P(ctx, tcn_ssl_ctxt_t *);
+    if (c == NULL) {
+        tcn_ThrowException(e, "ctx is null");
+        return;
+    }
+
+#if defined(TCN_NO_OCSP)
+    tcn_ThrowException(e, "OCSP stapling is not supported");
+    return;
+
+#elif !defined(OPENSSL_IS_BORINGSSL)
+    //
+    // The client and server use slightly different return values to signal
+    // error and success to OpenSSL. We're going to do something naughty to
+    // align OpenSSL's and BoringSSL's APIs and simply tell OpenSSL to use
+    // a stubbed callback function that is always saying that things are OK.
+    //
+    const int *arg = (client ? &OCSP_CLIENT_ACK : &OCSP_SERVER_ACK);
+    if (SSL_CTX_set_tlsext_status_arg(c->ctx, arg) <= 0L) {
+        tcn_ThrowException(e, "SSL_CTX_set_tlsext_status_arg() failed");
+        return;
+    }
+
+    if (SSL_CTX_set_tlsext_status_cb(c->ctx, openssl_ocsp_callback) <= 0L) {
+        tcn_ThrowException(e, "SSL_CTX_set_tlsext_status_cb() failed");
+        return;
+    }
+#endif
+}
+
+/**
+ * Disables OCSP stapling for the given SSLContext
+ */
+TCN_IMPLEMENT_CALL(void, SSLContext, disableOcsp)(TCN_STDARGS, jlong ctx) {
+    tcn_ssl_ctxt_t *c = J2P(ctx, tcn_ssl_ctxt_t *);
+    if (c == NULL) {
+        tcn_ThrowException(e, "ctx is null");
+        return;
+    }
+
+// This does nothing in BoringSSL
+#if !defined(TCN_NO_OCSP) && !defined(OPENSSL_IS_BORINGSSL)
+    SSL_CTX_set_tlsext_status_cb(c->ctx, NULL);
+    SSL_CTX_set_tlsext_status_arg(c->ctx, NULL);
+#endif
+}
+
+/**
+ * Enables OCSP stapling on the SSLEngine.
+ */
+TCN_IMPLEMENT_CALL(void, SSL, enableOcsp)(TCN_STDARGS, jlong ssl) {
```

Comment:
IMHO it would be nice to keep everything that is exposed via `SSL` java class in ssl.c and everything that is exposed via `SSLContext` in sslcontext.c . This makes it easier to find stuff. That said I am not strong on this...